Add  feature/plugin development guide (#435)
theforeman/foremanctl: docs/developer/how-to-add-a-feature.md

# How to Add a Feature

This guide covers what you need to do to add a feature to foremanctl, using Remote Execution (REX) as a running example.

## What is a feature?

> A feature in foremanctl extends the functionality of Foreman, Smart Proxy, and Hammer CLI by installing and configuring the relevant plugin packages in each component. The foremanctl tool automatically resolves dependencies, installs packages, and deploys configuration based on what you define in `src/features.yaml`.

Reviewer: ```suggestion
A feature in foremanctl extends the functionality of the deployment by installing and configuring the additional components. The foremanctl tool automatically resolves dependencies, installs packages, and deploys configuration based on what you define in `src/features.yaml`.
```

IOP is a great way to verify such statements, as it's a totally standalone thing, that gets deployed and then integrated with the rest of the ecosystem.

## Step 1: Register the Feature

A feature can belong to three components:

1. **Foreman plugin** — installed into the Foreman container
2. **Smart Proxy plugin** — installed into the foreman-proxy container
3. **Hammer CLI plugin** — installed on the host machine (Hammer is not containerized)

All features must be registered in the central feature registry.

**Feature Registry:** `src/features.yaml`

Add your feature definition to above file:

Example: Remote Execution Feature

```yaml
remote-execution:
  description: Remote Execution plugin for Foreman
  foreman:
    plugin_name: foreman_remote_execution
  foreman_proxy:
    plugin_name: remote_execution_ssh
  hammer: foreman_remote_execution
  dependencies:
    - dynflow
```

See [Feature Metadata Reference](feature-metadata.md) for the full list of properties.

### Naming

Plugin names must match what the ecosystem expects:

- `foreman.plugin_name` -- the plugin/gem name for foreman (e.g. `foreman_remote_execution`)

> - `foreman_proxy.plugin_name` -- the smart proxy plugin/gem name without `smart_proxy_` prefix (e.g. `smart_proxy_remote_execution_ssh`)

Reviewer: ```suggestion
- `foreman_proxy.plugin_name` -- the smart proxy plugin/gem name without `smart_proxy_` prefix (e.g. `remote_execution_ssh`)
```

> - `hammer` -- the suffix for the RPM package (e.g. `foreman_remote_execution` → `hammer-cli-plugin-foreman_remote_execution`)

Reviewer: strictly speaking it's not the suffix of the RPM package, it's the "plugin name" (aka: the name of the file in `cli.modules.d` minus the `.yml`) that we expose in packaging to make it easier to install things on different OSes. in most cases it's the "gem name without the hammer_cli_ prefix"

Author: yes, plugin name is better, but in case of gem name i saw couple of gems and the name of those was not consistent in using _ or - , like [hammer-cli-foreman-ansible](https://github.com/theforeman/hammer-cli-foreman-ansible) uses - but [hammer_cli_foreman_remote_execution](https://github.com/theforeman/hammer_cli_foreman_remote_execution)
uses _  

added `plugin/gem name without the `hammer_cli_` prefix`

Reviewer: ansible too uses underscores for the gem name -- only the repo is missnamed :)

```
% git grep -E 'global (gem|plugin).name' packages/*/rubygem-hammer*      
packages/foreman/rubygem-hammer_cli/rubygem-hammer_cli.spec:%global gem_name hammer_cli
packages/foreman/rubygem-hammer_cli_foreman/rubygem-hammer_cli_foreman.spec:%global gem_name hammer_cli_foreman
packages/foreman/rubygem-hammer_cli_foreman/rubygem-hammer_cli_foreman.spec:%global plugin_name foreman
packages/katello/rubygem-hammer_cli_foreman_rh_cloud/rubygem-hammer_cli_foreman_rh_cloud.spec:%global gem_name hammer_cli_foreman_rh_cloud
packages/katello/rubygem-hammer_cli_foreman_rh_cloud/rubygem-hammer_cli_foreman_rh_cloud.spec:%global plugin_name foreman_rh_cloud
packages/katello/rubygem-hammer_cli_foreman_scc_manager/rubygem-hammer_cli_foreman_scc_manager.spec:%global gem_name hammer_cli_foreman_scc_manager
packages/katello/rubygem-hammer_cli_foreman_scc_manager/rubygem-hammer_cli_foreman_scc_manager.spec:%global plugin_name foreman_scc_manager
packages/katello/rubygem-hammer_cli_foreman_virt_who_configure/rubygem-hammer_cli_foreman_virt_who_configure.spec:%global gem_name hammer_cli_foreman_virt_who_configure
packages/katello/rubygem-hammer_cli_foreman_virt_who_configure/rubygem-hammer_cli_foreman_virt_who_configure.spec:%global plugin_name foreman_virt_who_configure
packages/katello/rubygem-hammer_cli_katello/rubygem-hammer_cli_katello.spec:%global gem_name hammer_cli_katello
packages/katello/rubygem-hammer_cli_katello/rubygem-hammer_cli_katello.spec:%global plugin_name katello
packages/plugins/rubygem-hammer_cli_foreman_admin/rubygem-hammer_cli_foreman_admin.spec:%global gem_name hammer_cli_foreman_admin
packages/plugins/rubygem-hammer_cli_foreman_ansible/rubygem-hammer_cli_foreman_ansible.spec:%global gem_name hammer_cli_foreman_ansible
packages/plugins/rubygem-hammer_cli_foreman_ansible/rubygem-hammer_cli_foreman_ansible.spec:%global plugin_name foreman_ansible
packages/plugins/rubygem-hammer_cli_foreman_azure_rm/rubygem-hammer_cli_foreman_azure_rm.spec:%global gem_name hammer_cli_foreman_azure_rm
packages/plugins/rubygem-hammer_cli_foreman_azure_rm/rubygem-hammer_cli_foreman_azure_rm.spec:%global plugin_name foreman_azure_rm
packages/plugins/rubygem-hammer_cli_foreman_bootdisk/rubygem-hammer_cli_foreman_bootdisk.spec:%global gem_name hammer_cli_foreman_bootdisk
packages/plugins/rubygem-hammer_cli_foreman_bootdisk/rubygem-hammer_cli_foreman_bootdisk.spec:%global plugin_name foreman_bootdisk
packages/plugins/rubygem-hammer_cli_foreman_discovery/rubygem-hammer_cli_foreman_discovery.spec:%global gem_name hammer_cli_foreman_discovery
packages/plugins/rubygem-hammer_cli_foreman_discovery/rubygem-hammer_cli_foreman_discovery.spec:%global plugin_name foreman_discovery
packages/plugins/rubygem-hammer_cli_foreman_google/rubygem-hammer_cli_foreman_google.spec:%global gem_name hammer_cli_foreman_google
packages/plugins/rubygem-hammer_cli_foreman_google/rubygem-hammer_cli_foreman_google.spec:%global plugin_name foreman_google
packages/plugins/rubygem-hammer_cli_foreman_kubevirt/rubygem-hammer_cli_foreman_kubevirt.spec:%global gem_name hammer_cli_foreman_kubevirt
packages/plugins/rubygem-hammer_cli_foreman_kubevirt/rubygem-hammer_cli_foreman_kubevirt.spec:%global plugin_name foreman_kubevirt
packages/plugins/rubygem-hammer_cli_foreman_leapp/rubygem-hammer_cli_foreman_leapp.spec:%global gem_name hammer_cli_foreman_leapp
packages/plugins/rubygem-hammer_cli_foreman_leapp/rubygem-hammer_cli_foreman_leapp.spec:%global plugin_name foreman_leapp
packages/plugins/rubygem-hammer_cli_foreman_openscap/rubygem-hammer_cli_foreman_openscap.spec:%global gem_name hammer_cli_foreman_openscap
packages/plugins/rubygem-hammer_cli_foreman_openscap/rubygem-hammer_cli_foreman_openscap.spec:%global plugin_name foreman_openscap
packages/plugins/rubygem-hammer_cli_foreman_puppet/rubygem-hammer_cli_foreman_puppet.spec:%global gem_name hammer_cli_foreman_puppet
packages/plugins/rubygem-hammer_cli_foreman_puppet/rubygem-hammer_cli_foreman_puppet.spec:%global plugin_name foreman_puppet
packages/plugins/rubygem-hammer_cli_foreman_remote_execution/rubygem-hammer_cli_foreman_remote_execution.spec:%global gem_name hammer_cli_foreman_remote_execution
packages/plugins/rubygem-hammer_cli_foreman_remote_execution/rubygem-hammer_cli_foreman_remote_execution.spec:%global plugin_name foreman_remote_execution
packages/plugins/rubygem-hammer_cli_foreman_resource_quota/rubygem-hammer_cli_foreman_resource_quota.spec:%global gem_name hammer_cli_foreman_resource_quota
packages/plugins/rubygem-hammer_cli_foreman_resource_quota/rubygem-hammer_cli_foreman_resource_quota.spec:%global plugin_name foreman_resource_quota
packages/plugins/rubygem-hammer_cli_foreman_salt/rubygem-hammer_cli_foreman_salt.spec:%global gem_name hammer_cli_foreman_salt
packages/plugins/rubygem-hammer_cli_foreman_salt/rubygem-hammer_cli_foreman_salt.spec:%global plugin_name foreman_salt
packages/plugins/rubygem-hammer_cli_foreman_ssh/rubygem-hammer_cli_foreman_ssh.spec:%global gem_name hammer_cli_foreman_ssh
packages/plugins/rubygem-hammer_cli_foreman_ssh/rubygem-hammer_cli_foreman_ssh.spec:%global plugin_name foreman_ssh
packages/plugins/rubygem-hammer_cli_foreman_tasks/rubygem-hammer_cli_foreman_tasks.spec:%global gem_name hammer_cli_foreman_tasks
packages/plugins/rubygem-hammer_cli_foreman_tasks/rubygem-hammer_cli_foreman_tasks.spec:%global plugin_name foreman_tasks
packages/plugins/rubygem-hammer_cli_foreman_templates/rubygem-hammer_cli_foreman_templates.spec:%global gem_name hammer_cli_foreman_templates
packages/plugins/rubygem-hammer_cli_foreman_templates/rubygem-hammer_cli_foreman_templates.spec:%global plugin_name foreman_templates
packages/plugins/rubygem-hammer_cli_foreman_webhooks/rubygem-hammer_cli_foreman_webhooks.spec:%global gem_name hammer_cli_foreman_webhooks
packages/plugins/rubygem-hammer_cli_foreman_webhooks/rubygem-hammer_cli_foreman_webhooks.spec:%global plugin_name foreman_webhooks
```

this looks quite consistent to me :)

Author: ok thanks,  i was thinking gem name = repo name.
does `plugin/gem name without the hammer_cli_ prefix` work?

Reviewer: yes!

### Dependencies

Use `dependencies` to pull in other dependent features automatically(which are required for a feature to work). Mark these non-user facing features with `internal: true` to hide them from users:

> ```yaml

Reviewer: this example doesn't show a dependency being used, but `internal: true` being declared.

IMHO the `internal` part is not important for 99% of the features being developed and we should not put it so prominently in the docs (it is in the metadata spec)

Author: my reason to put it here in docs was because these docs are especially developer facing, and i think a dev should be aware about depedencies while implementing a feature.

`not important for 99% of the features`, you mean most plugins won't require dependencies in features.yml?

Reviewer: no, dependencies are perfectly fine and important here.

the fact that you can mark features as "internal" is not

Author: ok so if i understand correctly, `non-user facing features` misguides as thats not feature its dependency and should be stated as dependency

### Settings Template (required for Smart Proxy plugins)

Create a Jinja2 template at:

```
src/roles/foreman_proxy/templates/settings.d/<plugin_name>.yml.j2
```

The filename must exactly match `foreman_proxy.plugin_name`. The template must start with `:enabled: {{ feature_enabled }}` -- the system sets this to `"true"` or `"false"` automatically. Add any plugin-specific settings after that using Ruby symbol notation (`:key: value`).

Example: REX needs additional settings:

```yaml
---
:enabled: {{ feature_enabled }}
:ssh_identity_key_file: '~/.ssh/id_rsa_foreman_proxy'
:local_working_dir: '/var/tmp'
:remote_working_dir: '/var/tmp'
:socket_working_dir: '/var/tmp'
:mode: ssh
```

### Additional Tasks (optional)

If the plugin needs addtional setup beyond the settings file, create a feature specific task file at:

```
src/roles/foreman_proxy/tasks/feature/<plugin_name>.yaml
```

The filename must exactly match `foreman_proxy.plugin_name` with `.yaml` extension. These tasks only run when the feature is enabled. If the file doesn't exist, nothing happens.

> > **Note:** This works because [`feature.yaml`](../../src/roles/foreman_proxy/tasks/feature.yaml) uses `ansible.builtin.first_found` to look up `feature/<plugin_name>.yaml`, if the file is present, it gets included; if not, it's silently skipped.

Reviewer: implementation detail

# How to Add a Feature

This guide covers what you need to do to add a feature to foremanctl, using Remote Execution (REX) as a running example.

## What is a feature?

> A feature in foremanctl extends the functionality of Foreman, Smart Proxy, and Hammer CLI by installing and configuring the relevant plugin packages in each. The foremanctl tool automatically resolves dependencies, installs packages, and deploys configuration based on what you define in `src/features.yaml`.

Reviewer 2: ```suggestion
A feature in foremanctl extends the functionality of Foreman, Smart Proxy, and Hammer CLI by installing and configuring the relevant plugin packages in each component. The foremanctl tool automatically resolves dependencies, installs packages, and deploys configuration based on what you define in `src/features.yaml`.
```

## Prerequisites

Before adding a feature, you should:

- Know which services your plugin extends (does it add a Foreman plugin, a Smart Proxy plugin, a Hammer CLI plugin, or some combination?)
- Know what configuration the plugin needs (settings files, credentials, extra mounts, etc.)
- Be aware that Smart Proxy plugins only take effect when `foreman-proxy` is an enabled feature, and Hammer plugins only when `hammer` is enabled. Depending on the deployment's flavor, these may already be included, see [Deployment Design](deployment.md) for details.

## Step 1: Register the Feature

A feature can belong to three components:

1. **Foreman plugin** - Installed into the Foreman container
2. **Smart Proxy plugin** - Installed into foreman-proxy container

> 3. **Hammer CLI plugin** - Installed on host machine(as we currently don't ship hammer in container)

Reviewer 2: This is a duplicate Step 1. I suggest we move this list of components into the first bullet in Prereqs.

Would also be good to clarify the naming of services/components, right now the text uses both.

Author: adding to Prereqs feels odd, as this is more info than requirement. but thanks for pointing duplicate thing, will update. also added components for consistency

## Step 1: Register the Feature

All features must be registered in the central feature registry.

**File:** `src/features.yaml`

> Add your feature definition to this file:

Reviewer 2: ```suggestion
Register your feature by adding the feature definition to this file:

**File:** `src/features.yaml`

Use the following structure:
```
